In [230]:
# import libraries
import pandas as pd
import numpy as np
import os

In [231]:
#importing df_prods from the Original Data and orders_wrangled.csv as df_orders from Prepared Data
path=r'/Users/roxanamiu/20.03.2024_Instacart_Basket_Analysis'
df_prods=pd.read_csv(r'/Users/roxanamiu/20.03.2024_Instacart_Basket_Analysis/2.Data/Original_Data/products.csv',index_col=False)
df_ords=pd.read_csv(r'/Users/roxanamiu/20.03.2024_Instacart_Basket_Analysis/2.Data/Prepared_Data/orders_wrangled.csv',index_col=False)

In [232]:
df_test = pd.DataFrame()

In [233]:
df_test['mix'] = ['a', 'b', 1, True]

In [234]:
df_test.head()

mix
0     a
1     b
2     1
3  True

## Finding if all the data from a column have the same type and changing the type

In [235]:
df_test.dtypes

mix    object
dtype: object

In [236]:
#formula to check for mixed type of data inside the same column of a dataframe
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [237]:
print (weird)

0    False
1    False
2     True
3     True
dtype: bool


In [238]:
df_test['mix'] = df_test['mix'].astype('str')

In [239]:
df_test.dtypes

mix    object
dtype: object

In [240]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [241]:
print (weird)

0    False
1    False
2    False
3    False
dtype: bool


## Finding missing values

In [242]:
#if a cell contains null values then the boolean function adds up values of 1 
# the conclusion is that column product_name contains 16 empty cells
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [243]:
#creating a new data frame with the rows that contains the empty cells 
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [244]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [245]:
#formula for returning the number of column and rown in a dataframe
df_prods.shape

(49693, 5)

In [246]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [247]:
df_prods_clean.shape

(49677, 5)

## Handling duplicates


#finding duplicates:create a new df only with the duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [248]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [249]:
#checking the number of rows before removing duplicates
df_prods_clean.shape

(49677, 5)

In [250]:
df_prods_clean_nodups=df_prods_clean.drop_duplicates()

In [251]:
df_prods_clean_nodups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

## Exporting new clean data frame as products_checked.csv in Prepared Data

In [252]:
df_prods_clean_nodups.to_csv(os.path.join(path, '2.Data','Prepared_Data', 'products_checked.csv'),index=False)

## Exercise 4.5. Homework

## Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [253]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       hour_of_order  days_since_prior_order  
count   3.421083e+06            3.214874e+06  
mean    1.345202e+01            1.111484e+01  
std     4.226088e+00            9.206737e+00  
min     0.000000e+00            0.000000e+00  
25%     1.000000e+01            4.000000e+00  
50%     1.300000e+01            7.000000e+00  
75%     1.600000e+01            1.500000e+01  
max     2.300000e+01            3.000000e+01

In [254]:
#changing the type of the order_id, user_id and order_number to string

In [255]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [256]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [257]:
df_ords['order_number'] = df_ords['order_number'].astype('str')

In [258]:
df_ords.describe()

orders_day_of_week  hour_of_order  days_since_prior_order
count        3.421083e+06   3.421083e+06            3.214874e+06
mean         2.776219e+00   1.345202e+01            1.111484e+01
std          2.046829e+00   4.226088e+00            9.206737e+00
min          0.000000e+00   0.000000e+00            0.000000e+00
25%          1.000000e+00   1.000000e+01            4.000000e+00
50%          3.000000e+00   1.300000e+01            7.000000e+00
75%          5.000000e+00   1.600000e+01            1.500000e+01
max          6.000000e+00   2.300000e+01            3.000000e+01

In [259]:
df_ords.dtypes

order_id                   object
user_id                    object
order_number               object
orders_day_of_week          int64
hour_of_order               int64
days_since_prior_order    float64
dtype: object

In [260]:
df_ords.describe()

orders_day_of_week  hour_of_order  days_since_prior_order
count        3.421083e+06   3.421083e+06            3.214874e+06
mean         2.776219e+00   1.345202e+01            1.111484e+01
std          2.046829e+00   4.226088e+00            9.206737e+00
min          0.000000e+00   0.000000e+00            0.000000e+00
25%          1.000000e+00   1.000000e+01            4.000000e+00
50%          3.000000e+00   1.300000e+01            7.000000e+00
75%          5.000000e+00   1.600000e+01            1.500000e+01
max          6.000000e+00   2.300000e+01            3.000000e+01

In [261]:
# The count of the days_since_prior_order is smaller than the other rows, this means there are some informations missing

In [262]:
# the min and the max values are in norm, for the days of the week between 0 and 6 and for the hours between 0 and 23 

## Check for mixed-type data in your df_ords dataframe.

In [263]:
for col in df_ords.columns.tolist():
  weird1 = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird1]) > 0:
    print (col)

In [264]:
print (weird1)

0          False
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Length: 3421083, dtype: bool


# No mixt data types were found#

## Run a check for missing values in your df_ords dataframe.

In [265]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_of_order                  0
days_since_prior_order    206209
dtype: int64

# There are 206209 cells with no value under the days_since_prior_order column

In [266]:
## since there are many values missing, actually the logic answer is that for the first orders there are no days since prior order 
## this tells me we have to leave everything how it is

In [267]:
#checking my results
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_of_order                  0
days_since_prior_order    206209
dtype: int64

In [268]:
## Run a check for duplicate values in your df_ords data.
#In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [269]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [270]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, hour_of_order, days_since_prior_order]
Index: []

In [271]:
#no duplicates were found

## Exporting the df_prods and the df_ords as csv files as orders_cleaned.csv and products_cleaned.csv

In [272]:
df_ords.to_csv(os.path.join(path, '2.Data','Prepared_Data', 'orders_cleaned.csv'),index=False)
df_prods.to_csv(os.path.join(path, '2.Data','Prepared_Data', 'products_cleaned.csv'),index=False)